### ISSUE TITLE
A more robust method of parsing would be nice

### ISSUE THREAD

**Issue by esDotDev** [albertodev01/fraction#15]
I am trying to parse an incoming dataset, and I do not know of the values will be formatted as "1", "1 1/2", "1/2", it may also have extra spaces inside mixed-fractions.

Is there any way to parse this to a double without doing a bunch of string pre-processing myself? It would also be nice to handle things like extra-spaces automatically, and just throw a string at the library and get back a double., regardless of whether it's a fraction, mixed fraction, ignoring extra whitespace.

**Comment by albertodev01** [comment]
Hello, thank you for reaching out 😄

Would you be able to give me an example of String parsing that you find difficult?

> it may also have extra spaces inside mixed-fractions

This might be hard to parse. If you gave me the `-5   1/  3` string, how would I know it is a mixed fraction? If you removed the spaces, you would get `-51/3` which is wrong.

Or also, if you had `15   21/54` for example, how would you parse it? Trimming doesn't work because `1521/54` gives me no hints on where the space is. 

> It would also be nice to handle things like extra-spaces automatically

Could you please provide me an example?

**Comment by esDotDev** [comment]
So here's what I have to write now (I think), I need to manually look for '/' and also call trim() to remove any trailing spaces, or I get exceptions:
```
double? matchDouble;
if (matchValue.contains('/')) {
  if (matchValue.contains(' ')) {
    matchDouble = MixedFraction.fromString(fraction.trim()).toDouble(); // handle `1 1/2`
  } else {
    matchDouble = Fraction.fromString(fraction.trim()).toDouble(); handle `1/2`
  }
} else {
  matchDouble = double.tryParse(matchValue); handle `1`
}
if(matchDouble != null) ...
```
I would like to just write:
```
double? matchDouble = Fraction.fromString(fraction).toDouble(); // can handle ` 1 ', '1/2 ', ' 1 1/2'
if(matchDouble != null) ...
```

**Comment by esDotDev** [comment]
Thinking on this some more, there it would not make sense to expect Fraction.fromString to work with 1.0 or something, so maybe something like this would be ideal:
```dart
double? result;
if(Fraction.isFraction(value){
  result = Fraction.fromString(fraction).toDouble(); // can handle ` 1 ', '1/2 ', ' 1 1/2'
} else {
  result = double.tryParse(value); // handles ` 1.5 ` 
}
```

**Comment by albertodev01** [comment]
Hello!

Yes, I wouldn't change `fromString` constructors because, for example, having `Fraction.fromString` to accept everything (1, 1/1 and 1 1/1) doesn't sound good to me. I prefer keeping a single responsibility for each method rather than giving it too much features 🙂 

---

Right now I don't have much time but you gave me the idea to work on this feature! I will make `Fraction` and `MixedFraction` subtypes of a new class called `Rational` which would look like this:

```dart
abstract class Rational {
  final int numerator;
  final int denominator:

  const Rational({
    required this.numerator,
    required this.denominatr,
  });

  double toDouble();
  String toString();

  static Rational tryParse(String value) {
    // return either 'Fraction' or 'MixedFraction'
  }
}
```

In this way you could call `Rational.tryParse('1')`, `Rational.tryParse('1/1')` or `Rational.tryParse('1 1/1')` but it would always return either `Fraction` or `MixedFraction`. If you want a double, you can just call `toDouble` 🙂 

I don't have much time now but I've put this in my todo list!

**Comment by esDotDev** [comment]
Nice! Ya that sounds good, cpl other things that would be nice:
* Do a trim inside of the parse method (like double.tryParse does)
* Return null instead of throwing exception, easier to work with and closer to the native tryParse method
* Be more forgiving of extra spaces, like `1       1/2` should probably be a valid fraction?

Close it if you feel like it, I've already implemented the checks, so this is more for the next user.

fwiw, I used your lib to write some code that can pluck out fractions from a recipe, and multiple them, worked awesome, thanks for the assist!

https://user-images.githubusercontent.com/736973/152560463-9802a619-bf35-4fad-8bc0-572996ae5091.mp4

**Comment by albertodev01** [comment]
Thank you for sharing 💙 

The next release will include these parsing additions for sure!

**Comment by esDotDev** [comment]
fwiw, ideal code with the new proposal from usage perspective is like:
```
double? value = Rational.tryParse(string)?.toDouble() ?? double.tryParse(string);
if(value != null) ... 
```

**Comment by albertodev01** [comment]
@esDotDev I've published the [4.0.0](https://pub.dev/packages/fraction) version which includes some fixes and the update we were talking about 🙂 

```dart
// This is a 'Fraction' object
Rational.tryParse('1/5'); // 1/5

// This is a 'MixedFraction' object
Rational.tryParse('2 4/7'); // 2 4/7

// This is 'null' because the given string doesn't represent a fraction or a mixed fraction
Rational.tryParse(''); // null
```

**Comment by esDotDev** [comment]
Cool! ty!
